ext2/dir.c

```diff
@@ -129,9 +129,14 @@ int _ext2_dir_read(ext2_t *fs, ext2_obj_t *dir, offs_t offs, struct dirent *res,
 	if ((entry = (ext2_dirent_t *)malloc(len)) == NULL)
 		return -ENOMEM;
 
-	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) != len) {
+	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < 0) {
 		free(entry);
-		return (ret < 0) ? (int)ret : -ENOENT;
+		return ret;
+	}
+
+	if (((size_t)ret < sizeof(ext2_dirent_t)) || ((size_t)ret < sizeof(ext2_dirent_t) + entry->len)) {
```

Comment:
`((size_t)ret < sizeof(ext2_dirent_t))` condition is redundant (if `entry->len` >= 0)
```suggestion
	if ((size_t)ret < sizeof(ext2_dirent_t) + entry->len) {
```

Reply:
not really - if the first part is true You won't access garbage value `entry->len`.

Still - not readable enough (the check might be deleted during next refactoring) - eg. by @agkaminski  :P

IMHO would be better to:
```c
if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < (ssize_t)sizeof(ext2_dirent_t)) {
  // handle read error / partial read
}
// other checks accessing `entry` follow
```

Side note - to be MISRA-compliant the assignment should be done before the `if`

Reply:
> not really - if the first part is true You won't access garbage value `entry->len`.
> 
> Still - not readable enough (the check might be deleted during next refactoring) - eg. by @agkaminski :P
> 
> IMHO would be better to:
> 
> ```c
> if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < (ssize_t)sizeof(ext2_dirent_t)) {
>   // handle read error / partial read
> }
> // other checks accessing `entry` follow
> ```
> 
> Side note - to be MISRA-compliant the assignment should be done before the `if`

I've just moved the assignment before the `if` and split the condition.